code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/BaseConstraintsProvider.java

```diff
@@ -106,49 +110,59 @@ protected void addComputeConstraints(AllocationAttempt attempt, List<Constraint>
     protected void addStorageConstraints(AllocationAttempt attempt, List<Constraint> constraints) {
         for (Map.Entry<Volume, Set<StoragePool>> entry : attempt.getPools().entrySet()) {
             Volume volume = entry.getKey();
-            List<? extends VolumeStoragePoolMap> vspms = mapDao.findNonRemoved(VolumeStoragePoolMap.class, Volume.class, volume.getId());
-            boolean alreadyMappedToPool = vspms.size() > 0;
+            boolean alreadyMappedToPool = entry.getValue().size() > 0;
             VolumeValidStoragePoolConstraint volumeToPoolConstraint = new VolumeValidStoragePoolConstraint(volume, alreadyMappedToPool);
-            if (alreadyMappedToPool) {
```

Comment:
@ibuildthecloud I added all this alreadyMappedToPool in my last PR when fixing allocation bugs. But, I realized that my `mapDao.findNonRemoved(VolumeStoragePoolMap.class ...)` lookup was just duplicating what was already done here: https://github.com/rancher/cattle/blob/0c4066f9fd2652f99d29989c2a29065f0378c20e/code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java#L161
and that the only real change we needed was that if the pools set returned by entry.getValue() was not empty, make the constraint an "exact" match.

So, my changes in this method actually simplify the logic back to before my last fix but then add the additional constraints needed to properly allocate stand-alone volumes.